explorer/ast/value.h

```diff
@@ -95,6 +96,38 @@ class Value {
   const Kind kind_;
 };
 
+// Contains the result of the evaluation of an expression, including a value,
+// the original expression category, and an optional address if available.
+class ExpressionResult {
+ public:
+  ExpressionResult(Nonnull<const Value*> v, std::optional<Address> address,
```

Comment:
It is used in at least one place, where this avoids branching based on expression category by having that constructor (`PatternMatch`). Might want to revisit in the dependent PR, where it might not be needed anymore though.